src/ansi-c/goto-conversion/goto_clean_expr.cpp

```diff
@@ -16,9 +16,172 @@ Author: Daniel Kroening, kroening@kroening.com
+  // All bounded variables are local.
+  declared_symbols.insert(qex.variables().begin(), qex.variables().end());
+
+  exprt res = true_exprt();
+
+  // Visit the quantifeir body along `pathes`.
+  while(!pathes.empty())
+  {
+    auto &current_it = pathes.back();
+    auto &path_condition = path_conditions_and_value_maps.back().first;
+    auto &value_map = path_conditions_and_value_maps.back().second;
+    INVARIANT(
+      current_it != where.instructions.end(),
+      "Quantifier body must have a unique end expression.");
+
+    // Add all local-declared symbols into `declared_symbols`.
+    if(current_it->is_decl())
+    {
+      declared_symbols.insert(current_it->decl_symbol());
+    }
+    // ASSIGN lhs := rhs
+    // Add the replace lhr <- value_map(rhs) to the current value_map.
+    if(current_it->is_assign())
+    {
+      // Check that if lhs is a declared symbol.
+      auto lhs = current_it->assign_lhs();
+      INVARIANT(
+        declared_symbols.count(find_base_symbol(lhs)),
+        "quantifier must not contain side effects");
+      exprt rhs = current_it->assign_rhs();
+      INVARIANT(
+        !has_subexpr(rhs, ID_function_call),
+        "quantifier must not contain side effects");
+      value_map.replace(rhs);
+      value_map.set(to_symbol_expr(current_it->assign_lhs()), rhs);
+    }
+    // GOTO label
+    // -----------
+    // Move the current targett to label.
+    // or
+    // IF cond GOTO label
+    // ----------
+    // Move the current targett to targett+1 with path condition
+    // path_condition && !cond;
+    // and add a new path starting from label with path condition
+    // path_condition && cond.
+    if(current_it->is_goto())
+    {
+      if(current_it->has_condition())
+      {
+        auto next_it = current_it->targets.front();
+        exprt copy_path_condition = path_condition;
+        replace_symbolt copy_symbol_map = value_map;
+        path_condition =
+          and_exprt(path_condition, not_exprt(current_it->condition()));
+        path_conditions_and_value_maps.push_back(std::make_pair(
+          and_exprt(copy_path_condition, current_it->condition()),
+          copy_symbol_map));
+        current_it++;
+        pathes.push_back(next_it);
+      }
+      else
+      {
+        current_it = current_it->targets.front();
+      }
+      continue;
+    }
+    // EXPRESSION(expr)
+    // The last instruction is an exression statement.
+    // We add the predicate path_condition ==> value_map(expr) to res.
+    if(current_it == last)
+    {
+      exprt copy_of_last_expr = last_expr;
+      value_map.replace(copy_of_last_expr);
+      res = and_exprt(res, implies_exprt(path_condition, copy_of_last_expr));
+      path_conditions_and_value_maps.pop_back();
+      pathes.pop_back();
+      continue;
+    }
+    // DEAD symbol
+    // Remove symbol from declared_symbols.
+    if(current_it->is_dead())
+    {
+      declared_symbols.erase(current_it->decl_symbol());
+    }
```

Comment:
I don't think you need to bother: symbol names are unique, so even an assignment to a variable that has been marked dead still refers to that local variable and cannot possibly be about a side effect.